File: finished/go/accessibility-report/artifact.go
package main

import (
	"crypto/sha256"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

type ArtifactState struct {
	path        string
	existed     bool
	fingerprint [sha256.Size]byte
}

// These artifact helpers are for the optional HTML extension, not the core review.
func CaptureArtifactState(workingDirectory, fileName string) (ArtifactState, error) {
	if fileName == "" || fileName == "." || fileName == ".." || filepath.Base(fileName) != fileName {
		return ArtifactState{}, fmt.Errorf("Artifact name must be a single file name.")
	}
	directory, err := filepath.Abs(workingDirectory)
	if err != nil {
		return ArtifactState{}, err
	}
	info, err := os.Stat(directory)
	if err != nil {
		return ArtifactState{}, err
	}
	if !info.IsDir() {
		return ArtifactState{}, fmt.Errorf("Artifact working directory is not a directory.")
	}
	state := ArtifactState{path: filepath.Join(directory, fileName)}
	if _, err := os.Lstat(state.path); err != nil {
		if os.IsNotExist(err) {
			return state, nil
		}
		return ArtifactState{}, err
	}
	content, err := readArtifact(state.path)
	if err != nil {
		return ArtifactState{}, err
	}
	state.existed = true
	state.fingerprint = sha256.Sum256(content)
	return state, nil
}

func VerifyArtifactUpdate(state ArtifactState) error {
	content, err := readArtifact(state.path)
	if err != nil {
		return fmt.Errorf("No output update was verified. %w", err)
	}
	if len(content) == 0 {
		return fmt.Errorf("No output update was verified. The output is empty.")
	}
	if state.existed && sha256.Sum256(content) == state.fingerprint {
		return fmt.Errorf("No output update was verified.")
	}
	return nil
}

func readArtifact(path string) ([]byte, error) {
	before, err := os.Lstat(path)
	if err != nil {
		return nil, err
	}
	if !before.Mode().IsRegular() {
		return nil, fmt.Errorf("Artifact %s must be a regular, nonsymlink file.", path)
	}
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	content, readErr := io.ReadAll(file)
	closeErr := file.Close()
	if readErr != nil {
		return nil, readErr
	}
	if closeErr != nil {
		return nil, closeErr
	}
	after, err := os.Lstat(path)
	if err != nil {
		return nil, err
	}
	if !after.Mode().IsRegular() || !os.SameFile(before, after) || before.Size() != after.Size() || !before.ModTime().Equal(after.ModTime()) {
		return nil, fmt.Errorf("Artifact %s changed while being read.", path)
	}
	return content, nil
}

File: finished/rust/accessibility-report/src/artifact.rs
use std::hash::{DefaultHasher, Hash, Hasher};
use std::io;
use std::path::{Component, Path, PathBuf};

#[derive(Debug)]
pub struct ArtifactState {
    path: PathBuf,
    fingerprint: Option<(u64, usize)>,
}

// These artifact helpers are for the optional HTML extension, not the core review.
pub fn capture_artifact_state(
    working_directory: impl AsRef<Path>,
    file_name: &str,
) -> io::Result<ArtifactState> {
    if file_name.is_empty()
        || Path::new(file_name).components().count() != 1
        || !matches!(
            Path::new(file_name).components().next(),
            Some(Component::Normal(_))
        )
    {
        return Err(io::Error::other(
            "Artifact name must be a single file name.",
        ));
    }
    let directory = working_directory.as_ref().canonicalize()?;
    if !directory.is_dir() {
        return Err(io::Error::other(
            "Artifact working directory is not a directory.",
        ));
    }
    let path = directory.join(file_name);
    let fingerprint = match std::fs::symlink_metadata(&path) {
        Ok(_) => Some(artifact_fingerprint(&path)?),
        Err(error) if error.kind() == io::ErrorKind::NotFound => None,
        Err(error) => return Err(error),
    };
    Ok(ArtifactState { path, fingerprint })
}

pub fn verify_artifact_update(state: &ArtifactState) -> io::Result<()> {
    let fingerprint = artifact_fingerprint(&state.path).map_err(|error| {
        io::Error::new(
            error.kind(),
            format!("No output update was verified. {error}"),
        )
    })?;
    if fingerprint.1 == 0 {
        return Err(io::Error::other(
            "No output update was verified. The output is empty.",
        ));
    }
    if state.fingerprint == Some(fingerprint) {
        return Err(io::Error::other("No output update was verified."));
    }
    Ok(())
}

fn artifact_fingerprint(path: &Path) -> io::Result<(u64, usize)> {
    let before = std::fs::symlink_metadata(path)?;
    if !before.is_file() || before.file_type().is_symlink() {
        return Err(io::Error::other(
            "Artifact must be a regular, nonsymlink file.",
        ));
    }
    let content = std::fs::read(path)?;
    let after = std::fs::symlink_metadata(path)?;
    if !after.is_file()
        || after.file_type().is_symlink()
        || before.len() != after.len()
        || before.modified()? != after.modified()?
    {
        return Err(io::Error::other("Artifact changed while being read."));
    }
    let mut hasher = DefaultHasher::new();
    content.hash(&mut hasher);
    Ok((hasher.finish(), content.len()))
}
